go/vt/tabletserver/splitquery/equal_splits_algorithm.go

```diff
@@ -0,0 +1,226 @@
+	}
+	if splitParams.splitCount <= 0 {
+		return nil, fmt.Errorf("using the EQUAL_SPLITS algorithm in SplitQuery requires a positive"+
+			" splitParams.splitCount. Got: %v", splitParams.splitCount)
+	}
+	result := &EqualSplitsAlgorithm{
+		splitParams: splitParams,
+		sqlExecuter: sqlExecuter,
+
+		minMaxQuery: buildMinMaxQuery(splitParams),
+	}
+	return result, nil
+}
+
+func (a *EqualSplitsAlgorithm) generateBoundaries() ([]tuple, error) {
+	// generateBoundaries should work for a split_column whose type is integral
+	// (both signed and unsigned) as well as for floating point values.
+	// We perform the calculation of the boundaries using precise big.Rat arithmetic and only
+	// truncate the result in the end if necessary.
+	// We do this since using float64 arithmetic does not have enough precision:
+	// for example, if max=math.MaxUint64 and min=math.MaxUint64-1000 then float64(min)==float64(max).
+	// On the other hand, using integer arithmetic for the case where the split_column is integral
+	// (i.e., rounding (max-min)/split_count to an integer) may cause very dissimilar interval
+	// lengths or a large deviation between split_count and the number of query-parts actually
+	// returned (consider min=0, max=9.5*10^6, and split_count=10^6).
+	// Note(erez): We can probably get away with using big.Float with ~64 bits of precision which
+	// will likely be more efficient. However, we defer optimizing this code until we see if this
+	// is a bottle-neck.
+	minValue, maxValue, err := a.executeMinMaxQuery()
+	if err != nil {
+		return nil, err
+	}
+	// If the table is empty, minValue and maxValue will be NULL.
+	if (minValue.IsNull() && !maxValue.IsNull()) ||
+		!minValue.IsNull() && maxValue.IsNull() {
+		panic(fmt.Sprintf("minValue and maxValue must both be NULL or both be non-NULL."+
+			" minValue: %v, maxValue: %v, splitParams.sql: %v",
+			minValue, maxValue, a.splitParams.sql))
+	}
+	if minValue.IsNull() {
+		log.Infof("Splitting an empty table. splitParams.sql: %v. Query will not be split.",
+			a.splitParams.sql)
+		return []tuple{}, nil
+	}
+	min, err := valueToBigRat(minValue)
+	if err != nil {
+		panic(fmt.Sprintf("Failed to convert min to a big.Rat: %v, min: %+v", err, min))
+	}
+	max, err := valueToBigRat(maxValue)
+	if err != nil {
+		panic(fmt.Sprintf("Failed to convert max to a big.Rat: %v, max: %+v", err, max))
+	}
+	minCmpMax := min.Cmp(max)
+	if minCmpMax > 0 {
+		panic(fmt.Sprintf("max(splitColumn) < min(splitColumn): max:%v, min:%v", max, min))
+	}
+	if minCmpMax == 0 {
+		log.Infof("max(%v)=min(%v)=%v. splitParams.sql: %v. Query will not be split.",
+			a.splitParams.splitColumns[0],
+			a.splitParams.splitColumns[0],
+			min,
+			a.splitParams.sql)
+		return []tuple{}, nil
+	}
+	// subIntervalSize = (max - min) / a.splitParams.splitCount
+	subIntervalSize := new(big.Rat)
+	subIntervalSize.Sub(max, min)
+	subIntervalSize.Quo(subIntervalSize, new(big.Rat).SetInt64(a.splitParams.splitCount))
```

Comment:
Better if boundary := min at the start of the loop, then
boundary.Add(boundary, subIntervalSize) every pass through the loop.